observ/object_proxy.py

```diff
@@ -0,0 +1,254 @@
+from operator import xor
+
+from .dep import Dep
+from .object_utils import get_object_attrs
+from .proxy import TYPE_LOOKUP, Proxy, proxy
+from .proxy_db import proxy_db
+from .traps import ReadonlyError
+
+
+class ObjectProxyBase(Proxy):
+    def __getattribute__(self, name):
+        if name in Proxy.__slots__:
+            return super().__getattribute__(name)
+
+        target = self.__target__
+        target_attrs = get_object_attrs(target)
+        if name not in target_attrs:
+            return getattr(target, name)
+
+        if Dep.stack:
+            keydeps = proxy_db.attrs(self)["keydep"]
+            if name not in keydeps:
+                keydeps[name] = Dep()
+            keydeps[name].depend()
+        value = getattr(target, name)
+        if self.__shallow__:
+            return value
+        return proxy(value, readonly=self.__readonly__)
+
+    def __setattr__(self, name, value):
+        if name in Proxy.__slots__:
+            return super().__setattr__(name, value)
+
+        if self.__readonly__:
+            raise ReadonlyError()
+
+        target = self.__target__
+        target_attrs = get_object_attrs(target)
+        attrs = proxy_db.attrs(self)
+        is_new_keydep = name not in attrs["keydep"]
+        is_new_target_attr = name not in target_attrs
+
+        old_value = getattr(target, name, None) if not is_new_target_attr else None
+        retval = setattr(target, name, value)
+
+        if is_new_keydep:
+            # we have not determined earlier that this attr
+            # should be reactive, so let's determine it now
+            target_attrs_new = get_object_attrs(target)
+            if name not in target_attrs_new:
+                # the set attr is not stateful (e.g. someone
+                # is attaching a bound method)
+                # so no need to track this modification
+                # TODO: we could cache this
+                return retval
+
+        new_value = getattr(target, name, None)
```

Comment:
Why is this `getattr` here? Isn't the `retval` the same as `new_value`? Maybe I'm missing some kind of nuance.

Reply:
`setattr` doesn't return anything I think. We could do without `retval` perhaps.

Reply:
How would this work with property getters/setters that change the inputs? Maybe we actually do need the getattr.

Reply:
Yeah.. That occurred to me as well

Reply:
Reviewed the code another time, doesn't need to change